kmultiversx/src/kmultiversx/kdist/mx-semantics/esdt.md

```diff
@@ -21,47 +21,35 @@ module ESDT
  // -------------------------------------------------------------
     rule keyWithNonce(TOK, NONCE) => TOK +Bytes Int2Bytes(NONCE, BE, Unsigned)
 
-    syntax InternalCmd ::= transferESDT ( Bytes , Bytes , ESDTTransfer )
-                         | transferESDTs    ( Bytes , Bytes , List )
-                         | transferESDTsAux ( Bytes , Bytes , List )
-
-
-    rule <commands> transferESDTs(FROM, TO, L)
-                 => transferESDTsAux(FROM, TO, L) 
-                 ~> appendToOutAccount(TO, OutputTransfer(FROM, L))
-                    ...
-         </commands>
-         <instrs> (#waitCommands ~> _) #Or .K </instrs>
-
-    rule <commands> transferESDTsAux(_, _, .List) => .K ... </commands>
-         <instrs> (#waitCommands ~> _) #Or .K </instrs>
-
-    rule <commands> transferESDTsAux(FROM, TO, ListItem(T:ESDTTransfer) Ls) 
-                 => transferESDT(FROM, TO, T) 
-                 ~> transferESDTsAux(FROM, TO, Ls)
-                    ... 
-         </commands>
-         <instrs> (#waitCommands ~> _) #Or .K </instrs>
-
-    rule <commands> transferESDT(FROM, TO, esdtTransfer(TOKEN, VALUE, 0)) 
-                 => checkAccountExists(FROM)
-                 ~> checkAccountExists(TO)
-                 ~> checkESDTBalance(FROM, TOKEN, VALUE)
-                 ~> addToESDTBalance(FROM, TOKEN, 0 -Int VALUE, false)
-                 ~> addToESDTBalance(TO,   TOKEN, VALUE, true)
-                    ... 
-         </commands>
-         <instrs> (#waitCommands ~> _) #Or .K </instrs>
-
-    rule <commands> transferESDT(FROM, TO, esdtTransfer(TOKEN, VALUE, NONCE)) 
-                 => checkAccountExists(FROM)
-                 ~> checkAccountExists(TO)
-                 ~> checkESDTBalance(FROM, keyWithNonce(TOKEN, NONCE), VALUE)
-                 ~> moveNFTToDestination(FROM, TO, keyWithNonce(TOKEN, NONCE), VALUE)
-                 ~> removeEmptyNft(FROM, keyWithNonce(TOKEN, NONCE))
-                    ... 
-         </commands>
-         <instrs> (#waitCommands ~> _) #Or .K </instrs>
+
+    syntax K ::= transferESDTs(Bytes, Bytes, List)       [function, total]
+               | transferESDTsAux(Bytes, Bytes, List)    [function, total]
+ // -------------------------------------------------------------------
+    rule transferESDTs(FROM, TO, L)
+      => transferESDTsAux(FROM, TO, L) 
+      ~> appendToOutAccount(TO, OutputTransfer(FROM, L))
+
+    rule transferESDTsAux(FROM, TO, ListItem(T) Ts)
```

Comment:
I'm always a bit nervous about having owise rules when extra conditions may be added implicitly by k to the non-owise rules. This does not happen here, but, e.g., if `transferESDT` would take a subsort of Bytes as the first argument (this does not exist, but let's assume it does for this example), then the owise rule would not handle only the empty list case, as intended, but also the case when FROM is bytes, but not that subsort.

In this case, it is feasible to check that the owise case does, indeed, behave as expected, but I think that the code would be easier to follow if you either add sorts to the FROM and TO variables, or you write the rule below with an actual empty list instead of an owise attribute.